spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -0,0 +1,516 @@
+
+        return totalAssets;
+    }
+
+    //--------------------------------------------------------------------------
+    // External & Public read methods
+
+    /// @notice Fetches the vault's asset token balance.
+    /// @param token The address of the asset ERC-20 token held by the vault.
+    /// @return The token balance.
+    function assetBalance(IERC20Upgradeable token) external view returns (uint256) {
+        return isVaultAsset(token) ? token.balanceOf(address(this)) : 0;
+    }
+
+    /// @notice Checks if the given token is held by the vault.
+    /// @param token The address of a token to check.
+    function isVaultAsset(IERC20Upgradeable token) public view returns (bool) {
+        return token == underlying || _deployed.contains(address(token)) || address(perp) == address(token);
+    }
+
+    /// @notice Total count of deployed asset tokens held by the vault.
+    /// @return The count.
+    function deployedCount() public view returns (uint256) {
+        return _deployed.length();
+    }
+
+    /// @notice The token address from the deployed asset token list by index.
+    /// @param i The index of a token.
+    function deployedAt(uint256 i) public view returns (ITranche) {
+        return ITranche(_deployed.at(i));
+    }
+
+    /// @notice Total count of earned income tokens held by the vault.
+    /// @return The count.
+    function earnedCount() public pure returns (uint256) {
+        return 1;
+    }
+
+    /// @notice The token address from the earned income token list by index.
+    /// @param i The index of a token.
+    function earnedAt(uint256 i) public view returns (IERC20Upgradeable) {
+        if (i > 0) {
+            revert OutOfBounds();
+        }
+        return IERC20Upgradeable(perp);
+    }
+
+    //--------------------------------------------------------------------------
+    // Private write methods
+
+    /// @dev Deposits underlying balance into the provided bond and receives tranche tokens in return.
+    function _tranche(IBondController bond) private returns (TrancheData memory) {
+        // Get bond's tranche data
+        TrancheData memory td = bond.getTrancheData();
+
+        // Get underlying balance
+        uint256 balance = underlying.balanceOf(address(this));
+
+        // Ensure initial deposit remains unspent
+        if (balance == 0) {
+            return td;
+        }
+
+        // balance is tranched
+        _checkAndApproveMax(underlying, address(bond), balance);
+        bond.deposit(balance);
+
+        // sync holdings
+        for (uint8 i = 0; i < td.trancheCount; i++) {
+            _syncDeployedAsset(td.tranches[i]);
+        }
+        _syncAsset(underlying);
+
+        return td;
+    }
+
+    /// @dev Rolls over freshly tranched tokens from the given bond for older tranches (close to maturity) from perp.
+    /// @return The amount of perps rolled over.
+    function _rollover(IPerpetualTranche perp_, TrancheData memory td) private returns (uint256) {
```

Comment:
We could, however the `getTrancheData` call is expensive and its cheaper to pass td to `_rollover` than query the storage again. Agree it would not be part of the vault interface